File: tests/dotnet/EncryptionTest.cs
/*
 * EncryptionTest.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.RegularExpressions;
using Xunit;

namespace PDFjet.NET {
/// <summary>AES-256 encryption with the standard security handler, read back with the passwords.</summary>
public class EncryptionTest {
    private static byte[] Encrypted(Compliance compliance, Passwords passwords, Permissions permissions) {
        MemoryStream stream = new MemoryStream();
        PDF pdf = new PDF(stream, compliance);
        pdf.SetTitle("Encrypted title");
        pdf.SetEncryption(new Encryption(pdf, passwords, permissions));
        Page page = new Page(pdf, Letter.PORTRAIT);
        new TextLine(new Font(pdf, CoreFont.HELVETICA), "Secret text").SetLocation(50f, 50f).DrawOn(page);
        pdf.Complete();
        return stream.ToArray();
    }

    private static byte[] Encrypted(string user, string owner) {
        return Encrypted(Compliance.PDF_1_7,
                new Passwords().SetUserPassword(user).SetOwnerPassword(owner),
                new Permissions().Grant(UserAccess.PRINT));
    }

    private static int AccessValue(byte[] pdf) {
        Match match = Regex.Match(TestSupport.Latin1(pdf), "/P (-?\\d+)");
        Assert.True(match.Success);
        return int.Parse(match.Groups[1].Value);
    }

    private static string Title(List<PDFobj> objects) {
        return TestSupport.Utf16Hex(TestSupport.FindObject(objects, "/Producer").GetValue("/Title"));
    }

    [Fact]
    public void ReadsBackWithTheUserOrTheOwnerPassword() {
        byte[] pdf = Encrypted("hello", "world");
        foreach (string password in new[] {"hello", "world"}) {
            List<PDFobj> objects = TestSupport.Read(pdf, password);
            Assert.Single(new PDF().GetPageObjects(objects));
            Assert.Equal("Encrypted title", Title(objects));
        }
    }

    [Fact]
    public void TheTitleIsNotInTheFileInPlainText() {
        string raw = TestSupport.Latin1(Encrypted("hello", "world"));
        Assert.DoesNotContain("feff0045006e0063", raw);
        Assert.Contains("/Encrypt ", raw);
    }

    [Fact]
    public void AWrongOrMissingPasswordFailsWithAMessage() {
        byte[] pdf = Encrypted("hello", "world");
        Exception wrong = Assert.ThrowsAny<Exception>(() => TestSupport.Read(pdf, "wrong"));
        Assert.Equal("The password of the PDF is not correct.", wrong.Message);
        Exception missing = Assert.ThrowsAny<Exception>(() => TestSupport.Read(pdf));
        Assert.Equal("The PDF needs a password.", missing.Message);
    }

    [Fact]
    public void AnEmptyUserPasswordOpensWithoutAPassword() {
        Assert.Equal("Encrypted title", Title(TestSupport.Read(Encrypted("", "owner"))));
    }

    [Fact]
    public void ANonAsciiPasswordIsUtf8() {
        byte[] pdf = Encrypted("пароль", "owner");
        Assert.Equal("Encrypted title", Title(TestSupport.Read(pdf, "пароль")));
    }

    [Fact]
    public void PasswordsAreCutAt127Bytes() {
        string password = new string('a', 200);
        byte[] pdf = Encrypted(password, "owner");
        Assert.Equal("Encrypted title", Title(TestSupport.Read(pdf, password.Substring(0, 127))));
        Assert.ThrowsAny<Exception>(() => TestSupport.Read(pdf, password.Substring(0, 126)));
    }

    [Fact]
    public void PermissionsAreANegativeNumberWithTheReservedBitsSet() {
        Assert.Equal(-3900, AccessValue(Encrypted("hello", "world")));
    }

    [Fact]
    public void PdfUaGrantsExtractionForAccessibility() {
        byte[] pdf = Encrypted(Compliance.PDF_UA_1,
                new Passwords().SetUserPassword("hello").SetOwnerPassword("world"),
                new Permissions().Grant(UserAccess.PRINT));
        int access = AccessValue(pdf);
        Assert.Equal(-3388, access);
        Assert.True(((UserAccess) access).HasFlag(UserAccess.EXTRACT_CONTENTS_FOR_ACCESSIBILITY));
    }

    [Fact]
    public void ACryptFilterThatIsAnObjectOfItsOwnIsFollowed() {
        // The /CF dictionary and the filter in it as objects of their own, as
        // pdf.js tests them in issue7665: the method was not found, and the
        // streams and the strings of the PDF were left encrypted.
        Func<int, string, PDFobj> obj = (number, raw) => {
            PDFobj o = new PDF().GetObject(System.Text.Encoding.Latin1.GetBytes(raw), 0);
            o.number = number;
            return o;
        };
        List<PDFobj> objects = new List<PDFobj> {
            obj(7, "7 0 obj << /StdCF 8 0 R >> endobj"),
            obj(8, "8 0 obj << /AuthEvent /DocOpen /CFM /AESV3 /Length 32 >> endobj"),
            obj(9, "9 0 obj << /StdCF << /CFM /AESV2 >> >> endobj"),
        };
        var cases = new (string cf, int want)[] {
            ("7 0 R", Decryptor.AES_256),
            ("9 0 R", Decryptor.AES_128),
            ("<< /StdCF 8 0 R >>", Decryptor.AES_256),
            ("<< /StdCF << /CFM /V2 >> >>", Decryptor.RC4),
            ("99 0 R", Decryptor.NONE),
        };
        foreach (var c in cases) {
            PDFobj encrypt = obj(6, "6 0 obj << /Filter /Standard /V 5 /CF " + c.cf + " /StmF /StdCF >> endobj");
            Assert.Equal((c.cf, c.want), (c.cf, Decryptor.GetMethod(encrypt, "/StdCF", objects)));
        }
    }
}
}

File: tests/java/com/pdfjet/encryption/EncryptionTest.java
/*
 * EncryptionTest.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet.encryption;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.pdfjet.Compliance;
import com.pdfjet.CoreFont;
import com.pdfjet.Encryption;
import com.pdfjet.Font;
import com.pdfjet.Letter;
import com.pdfjet.PDF;
import com.pdfjet.PDFobj;
import com.pdfjet.Page;
import com.pdfjet.TestSupport;
import com.pdfjet.TextLine;
import java.io.ByteArrayOutputStream;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.junit.jupiter.api.Test;

/** AES-256 encryption with the standard security handler, read back with the passwords. */
class EncryptionTest {
    private static byte[] encrypted(Compliance compliance, Passwords passwords, Permissions permissions) throws Exception {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        PDF pdf = new PDF(bos, compliance);
        pdf.setTitle("Encrypted title");
        pdf.setEncryption(new Encryption(pdf, passwords, permissions));
        Page page = new Page(pdf, Letter.PORTRAIT);
        new TextLine(new Font(pdf, CoreFont.HELVETICA), "Secret text").setLocation(50f, 50f).drawOn(page);
        pdf.complete();
        return bos.toByteArray();
    }

    private static byte[] encrypted(String user, String owner) throws Exception {
        return encrypted(Compliance.PDF_1_7,
                new Passwords().setUserPassword(user).setOwnerPassword(owner),
                new Permissions().grant(UserAccess.PRINT.getValue()));
    }

    private static int accessValue(byte[] pdf) {
        Matcher matcher = Pattern.compile("/P (-?\\d+)").matcher(TestSupport.latin1(pdf));
        assertTrue(matcher.find());
        return Integer.parseInt(matcher.group(1));
    }

    private static String title(List<PDFobj> objects) {
        return TestSupport.utf16Hex(TestSupport.findObject(objects, "/Producer").getValue("/Title"));
    }

    @Test
    void readsBackWithTheUserOrTheOwnerPassword() throws Exception {
        byte[] pdf = encrypted("hello", "world");
        for (String password : new String[] {"hello", "world"}) {
            List<PDFobj> objects = TestSupport.read(pdf, password);
            assertEquals(1, new PDF().getPageObjects(objects).size());
            assertEquals("Encrypted title", title(objects));
        }
    }

    @Test
    void theTitleIsNotInTheFileInPlainText() throws Exception {
        String raw = TestSupport.latin1(encrypted("hello", "world"));
        assertFalse(raw.contains("feff0045006e0063"));
        assertTrue(raw.contains("/Encrypt "));
    }

    @Test
    void aWrongOrMissingPasswordFailsWithAMessage() throws Exception {
        final byte[] pdf = encrypted("hello", "world");
        Exception wrong = assertThrows(Exception.class, () -> TestSupport.read(pdf, "wrong"));
        assertEquals("The password of the PDF is not correct.", wrong.getMessage());
        Exception missing = assertThrows(Exception.class, () -> TestSupport.read(pdf));
        assertEquals("The PDF needs a password.", missing.getMessage());
    }

    @Test
    void anEmptyUserPasswordOpensWithoutAPassword() throws Exception {
        List<PDFobj> objects = TestSupport.read(encrypted("", "owner"));
        assertEquals("Encrypted title", title(objects));
    }

    @Test
    void aNonAsciiPasswordIsUtf8() throws Exception {
        byte[] pdf = encrypted("пароль", "owner");
        assertEquals("Encrypted title", title(TestSupport.read(pdf, "пароль")));
    }

    @Test
    void passwordsAreCutAt127Bytes() throws Exception {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 200; i++) {
            sb.append('a');
        }
        final byte[] pdf = encrypted(sb.toString(), "owner");
        assertEquals("Encrypted title", title(TestSupport.read(pdf, sb.substring(0, 127))));
        assertThrows(Exception.class, () -> TestSupport.read(pdf, sb.substring(0, 126)));
    }

    @Test
    void permissionsAreANegativeNumberWithTheReservedBitsSet() throws Exception {
        assertEquals(-3900, accessValue(encrypted("hello", "world")));
    }

    @Test
    void pdfUaGrantsExtractionForAccessibility() throws Exception {
        byte[] pdf = encrypted(Compliance.PDF_UA_1,
                new Passwords().setUserPassword("hello").setOwnerPassword("world"),
                new Permissions().grant(UserAccess.PRINT.getValue()));
        int access = accessValue(pdf);
        assertEquals(-3388, access);
        assertTrue(UserAccess.EXTRACT_CONTENTS_FOR_ACCESSIBILITY.isSetIn(access));
    }
}
